// @xhs/hulk-env/dist/env.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNodeVersion = exports.getPodIP = exports.getEnvTag = exports.getHostname = exports.getServiceName = exports.getAppVersion = exports.getAppId = exports.getNamespace = exports.getZone = exports.getRegion = exports.isDebug = exports.isLocal = exports.isCi = exports.isSit = exports.isStaging = exports.isProd = exports.getEnv = void 0;
const pkg_1 = require("./pkg");
const ip_1 = require("./ip");
const enum_1 = require("./enum");
function getEnv() {
    switch (process.env.XHS_ENV) {
        case 'prod':
            return enum_1.Env.PROD;
        case 'staging':
            return enum_1.Env.STAGING;
        case 'sit':
            return enum_1.Env.SIT;
        case 'ci':
            return enum_1.Env.CI;
        default:
            return enum_1.Env.LOCAL;
    }
}
exports.getEnv = getEnv;
function isProd() {
    return getEnv() === enum_1.Env.PROD;
}
exports.isProd = isProd;
function isStaging() {
    return getEnv() === enum_1.Env.STAGING;
}
exports.isStaging = isStaging;
function isSit() {
    return getEnv() === enum_1.Env.SIT;
}
exports.isSit = isSit;
function isCi() {
    return getEnv() === enum_1.Env.CI;
}
exports.isCi = isCi;
function isLocal() {
    return getEnv() === enum_1.Env.LOCAL;
}
exports.isLocal = isLocal;
function isDebug() {
    return Boolean(process.env.DEBUG);
}
exports.isDebug = isDebug;
function getRegion() {
    return process.env.XHS_REGION;
}
exports.getRegion = getRegion;
function getZone() {
    return process.env.XHS_ZONE;
}
exports.getZone = getZone;
function getNamespace() {
    return process.env.NAMESPACE || 'frontend';
}
exports.getNamespace = getNamespace;
function getAppId() {
    return process.env.APPID || (0, pkg_1.getPKGInfo)()?.name;
}
exports.getAppId = getAppId;
function getAppVersion() {
    return (0, pkg_1.getPKGInfo)()?.version;
}
exports.getAppVersion = getAppVersion;
function getServiceName() {
    return process.env.XHS_SERVICE;
}
exports.getServiceName = getServiceName;
function getHostname() {
    return process.env.HOSTNAME;
}
exports.getHostname = getHostname;
function getEnvTag() {
    return process.env.SERVICE_ENV_TAG;
}
exports.getEnvTag = getEnvTag;
function getPodIP() {
    return process.env.POD_IP || (0, ip_1.getIPv4)();
}
exports.getPodIP = getPodIP;
function getNodeVersion() {
    return process.env.NODE_VERSION;
}
exports.getNodeVersion = getNodeVersion;
