// @xhs/hulk-xds/dist/eds/generated/envoy/api/v2/eds_connect.js
"use strict";
// @generated by protoc-gen-connect-es v0.8.5 with parameter "target=ts,import_extension=none"
// @generated from file envoy/api/v2/eds.proto (package envoy.api.v2, syntax proto3)
/* eslint-disable */
// @ts-nocheck
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndpointDiscoveryService = void 0;
const discovery_pb_1 = require("./discovery_pb");
const protobuf_1 = require("@bufbuild/protobuf");
/**
 * @generated from service envoy.api.v2.EndpointDiscoveryService
 */
exports.EndpointDiscoveryService = {
    typeName: "envoy.api.v2.EndpointDiscoveryService",
    methods: {
        /**
         * The resource_names field in DiscoveryRequest specifies a list of clusters
         * to subscribe to updates for.
         *
         * @generated from rpc envoy.api.v2.EndpointDiscoveryService.StreamEndpoints
         */
        streamEndpoints: {
            name: "StreamEndpoints",
            I: discovery_pb_1.DiscoveryRequest,
            O: discovery_pb_1.DiscoveryResponse,
            kind: protobuf_1.MethodKind.BiDiStreaming,
        },
        /**
         * @generated from rpc envoy.api.v2.EndpointDiscoveryService.DeltaEndpoints
         */
        deltaEndpoints: {
            name: "DeltaEndpoints",
            I: discovery_pb_1.DeltaDiscoveryRequest,
            O: discovery_pb_1.DeltaDiscoveryResponse,
            kind: protobuf_1.MethodKind.BiDiStreaming,
        },
        /**
         * @generated from rpc envoy.api.v2.EndpointDiscoveryService.FetchEndpoints
         */
        fetchEndpoints: {
            name: "FetchEndpoints",
            I: discovery_pb_1.DiscoveryRequest,
            O: discovery_pb_1.DiscoveryResponse,
            kind: protobuf_1.MethodKind.Unary,
        },
    }
};
