// @bufbuild/connect/dist/cjs/protocol-grpc/transport.js
"use strict";
// Copyright 2021-2023 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTransport = void 0;
const index_js_1 = require("../index.js");
const index_js_2 = require("../protocol/index.js");
const request_header_js_1 = require("./request-header.js");
const validate_response_js_1 = require("./validate-response.js");
const validate_trailer_js_1 = require("./validate-trailer.js");
/**
 * Create a Transport for the gRPC protocol.
 */
function createTransport(opt) {
    return {
        async unary(service, method, signal, timeoutMs, header, input) {
            const serialization = (0, index_js_2.createMethodSerializationLookup)(method, opt.binaryOptions, opt.jsonOptions, opt);
            return await (0, index_js_1.runUnary)({
                stream: false,
                service,
                method,
                url: (0, index_js_2.createMethodUrl)(opt.baseUrl, service, method),
                init: {},
                header: (0, request_header_js_1.requestHeaderWithCompression)(opt.useBinaryFormat, timeoutMs, header, opt.acceptCompression, opt.sendCompression),
                message: input instanceof method.I ? input : new method.I(input),
                signal: signal !== null && signal !== void 0 ? signal : new AbortController().signal,
            }, async (req) => {
                const uRes = await opt.httpClient({
                    url: req.url,
                    method: "POST",
                    header: req.header,
                    signal: req.signal,
                    body: (0, index_js_2.pipe)((0, index_js_2.createAsyncIterable)([req.message]), (0, index_js_2.transformSerializeEnvelope)(serialization.getI(opt.useBinaryFormat)), (0, index_js_2.transformCompressEnvelope)(opt.sendCompression, opt.compressMinBytes), (0, index_js_2.transformJoinEnvelopes)(), {
                        propagateDownStreamError: true,
                    }),
                });
                const { compression } = (0, validate_response_js_1.validateResponseWithCompression)(opt.useBinaryFormat, opt.acceptCompression, uRes.status, uRes.header);
                const message = await (0, index_js_2.pipeTo)(uRes.body, (0, index_js_2.transformSplitEnvelope)(opt.readMaxBytes), (0, index_js_2.transformDecompressEnvelope)(compression !== null && compression !== void 0 ? compression : null, opt.readMaxBytes), (0, index_js_2.transformParseEnvelope)(serialization.getO(opt.useBinaryFormat)), async (iterable) => {
                    var _a, e_1, _b, _c;
                    let message;
                    try {
                        for (var _d = true, iterable_1 = __asyncValues(iterable), iterable_1_1; iterable_1_1 = await iterable_1.next(), _a = iterable_1_1.done, !_a;) {
                            _c = iterable_1_1.value;
                            _d = false;
                            try {
                                const chunk = _c;
                                if (message !== undefined) {
                                    throw new index_js_1.ConnectError("protocol error: received extra output message for unary method", index_js_1.Code.InvalidArgument);
                                }
                                message = chunk;
                            }
                            finally {
                                _d = true;
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (!_d && !_a && (_b = iterable_1.return)) await _b.call(iterable_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                    return message;
                }, { propagateDownStreamError: false });
                (0, validate_trailer_js_1.validateTrailer)(uRes.trailer);
                if (message === undefined) {
                    throw new index_js_1.ConnectError("protocol error: missing output message for unary method", index_js_1.Code.InvalidArgument);
                }
                return {
                    stream: false,
                    service,
                    method,
                    header: uRes.header,
                    message,
                    trailer: uRes.trailer,
                };
            }, opt.interceptors);
        },
        async stream(service, method, signal, timeoutMs, header, input) {
            const serialization = (0, index_js_2.createMethodSerializationLookup)(method, opt.binaryOptions, opt.jsonOptions, opt);
            return (0, index_js_1.runStreaming)({
                stream: true,
                service,
                method,
                url: (0, index_js_2.createMethodUrl)(opt.baseUrl, service, method),
                init: {},
                signal: signal !== null && signal !== void 0 ? signal : new AbortController().signal,
                header: (0, request_header_js_1.requestHeaderWithCompression)(opt.useBinaryFormat, timeoutMs, header, opt.acceptCompression, opt.sendCompression),
                message: (0, index_js_2.pipe)(input, (0, index_js_2.transformNormalizeMessage)(method.I), {
                    propagateDownStreamError: true,
                }),
            }, async (req) => {
                const uRes = await opt.httpClient({
                    url: req.url,
                    method: "POST",
                    header: req.header,
                    signal: req.signal,
                    body: (0, index_js_2.pipe)(req.message, (0, index_js_2.transformNormalizeMessage)(method.I), (0, index_js_2.transformSerializeEnvelope)(serialization.getI(opt.useBinaryFormat)), (0, index_js_2.transformCompressEnvelope)(opt.sendCompression, opt.compressMinBytes), (0, index_js_2.transformJoinEnvelopes)(), { propagateDownStreamError: true }),
                });
                const { compression, foundStatus } = (0, validate_response_js_1.validateResponseWithCompression)(opt.useBinaryFormat, opt.acceptCompression, uRes.status, uRes.header);
                const res = Object.assign(Object.assign({}, req), { header: uRes.header, trailer: uRes.trailer, message: (0, index_js_2.pipe)(uRes.body, (0, index_js_2.transformSplitEnvelope)(opt.readMaxBytes), (0, index_js_2.transformDecompressEnvelope)(compression !== null && compression !== void 0 ? compression : null, opt.readMaxBytes), (0, index_js_2.transformParseEnvelope)(serialization.getO(opt.useBinaryFormat)), function (iterable) {
                        return __asyncGenerator(this, arguments, function* () {
                            yield __await(yield* __asyncDelegator(__asyncValues(iterable)));
                            if (!foundStatus) {
                                (0, validate_trailer_js_1.validateTrailer)(uRes.trailer);
                            }
                        });
                    }, { propagateDownStreamError: true }) });
                return res;
            }, opt.interceptors);
        },
    };
}
exports.createTransport = createTransport;
