// @xhs/hulk-xds/dist/eds/generated/udpa/annotations/status_pb.js
"use strict";
// THIS FILE IS DEPRECATED
// Users should instead use the corresponding proto in the xds tree.
// No new changes will be accepted here.
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusAnnotation = exports.PackageVersionStatus = void 0;
const protobuf_1 = require("@bufbuild/protobuf");
/**
 * @generated from enum udpa.annotations.PackageVersionStatus
 */
var PackageVersionStatus;
(function (PackageVersionStatus) {
    /**
     * Unknown package version status.
     *
     * @generated from enum value: UNKNOWN = 0;
     */
    PackageVersionStatus[PackageVersionStatus["UNKNOWN"] = 0] = "UNKNOWN";
    /**
     * This version of the package is frozen.
     *
     * @generated from enum value: FROZEN = 1;
     */
    PackageVersionStatus[PackageVersionStatus["FROZEN"] = 1] = "FROZEN";
    /**
     * This version of the package is the active development version.
     *
     * @generated from enum value: ACTIVE = 2;
     */
    PackageVersionStatus[PackageVersionStatus["ACTIVE"] = 2] = "ACTIVE";
    /**
     * This version of the package is the candidate for the next major version. It
     * is typically machine generated from the active development version.
     *
     * @generated from enum value: NEXT_MAJOR_VERSION_CANDIDATE = 3;
     */
    PackageVersionStatus[PackageVersionStatus["NEXT_MAJOR_VERSION_CANDIDATE"] = 3] = "NEXT_MAJOR_VERSION_CANDIDATE";
})(PackageVersionStatus || (exports.PackageVersionStatus = PackageVersionStatus = {}));
// Retrieve enum metadata with: proto3.getEnumType(PackageVersionStatus)
protobuf_1.proto3.util.setEnumType(PackageVersionStatus, "udpa.annotations.PackageVersionStatus", [
    { no: 0, name: "UNKNOWN" },
    { no: 1, name: "FROZEN" },
    { no: 2, name: "ACTIVE" },
    { no: 3, name: "NEXT_MAJOR_VERSION_CANDIDATE" },
]);
/**
 * @generated from message udpa.annotations.StatusAnnotation
 */
class StatusAnnotation extends protobuf_1.Message {
    /**
     * The entity is work-in-progress and subject to breaking changes.
     *
     * @generated from field: bool work_in_progress = 1;
     */
    workInProgress = false;
    /**
     * The entity belongs to a package with the given version status.
     *
     * @generated from field: udpa.annotations.PackageVersionStatus package_version_status = 2;
     */
    packageVersionStatus = PackageVersionStatus.UNKNOWN;
    constructor(data) {
        super();
        protobuf_1.proto3.util.initPartial(data, this);
    }
    static runtime = protobuf_1.proto3;
    static typeName = "udpa.annotations.StatusAnnotation";
    static fields = protobuf_1.proto3.util.newFieldList(() => [
        { no: 1, name: "work_in_progress", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
        { no: 2, name: "package_version_status", kind: "enum", T: protobuf_1.proto3.getEnumType(PackageVersionStatus) },
    ]);
    static fromBinary(bytes, options) {
        return new StatusAnnotation().fromBinary(bytes, options);
    }
    static fromJson(jsonValue, options) {
        return new StatusAnnotation().fromJson(jsonValue, options);
    }
    static fromJsonString(jsonString, options) {
        return new StatusAnnotation().fromJsonString(jsonString, options);
    }
    static equals(a, b) {
        return protobuf_1.proto3.util.equals(StatusAnnotation, a, b);
    }
}
exports.StatusAnnotation = StatusAnnotation;
