// @xhs/hulk-env/dist/esm/env.js
import { getPKGInfo } from './pkg';
import { getIPv4 } from './ip';
import { Env } from './enum';
export function getEnv() {
    switch (process.env.XHS_ENV) {
        case 'prod':
            return Env.PROD;
        case 'staging':
            return Env.STAGING;
        case 'sit':
            return Env.SIT;
        case 'ci':
            return Env.CI;
        default:
            return Env.LOCAL;
    }
}
export function isProd() {
    return getEnv() === Env.PROD;
}
export function isStaging() {
    return getEnv() === Env.STAGING;
}
export function isSit() {
    return getEnv() === Env.SIT;
}
export function isCi() {
    return getEnv() === Env.CI;
}
export function isLocal() {
    return getEnv() === Env.LOCAL;
}
export function isDebug() {
    return Boolean(process.env.DEBUG);
}
export function getRegion() {
    return process.env.XHS_REGION;
}
export function getZone() {
    return process.env.XHS_ZONE;
}
export function getNamespace() {
    return process.env.NAMESPACE || 'frontend';
}
export function getAppId() {
    return process.env.APPID || getPKGInfo()?.name;
}
export function getAppVersion() {
    return getPKGInfo()?.version;
}
export function getServiceName() {
    return process.env.XHS_SERVICE;
}
export function getHostname() {
    return process.env.HOSTNAME;
}
export function getEnvTag() {
    return process.env.SERVICE_ENV_TAG;
}
export function getPodIP() {
    return process.env.POD_IP || getIPv4();
}
export function getNodeVersion() {
    return process.env.NODE_VERSION;
}
